### ISSUE TITLE
Mark PrefetchCount as internal

### ISSUE THREAD

**Issue by JoshLove-msft** [Azure/azure-sdk-for-net#13077]
We are planning on leaving out Prefetch for GA:

There are quite a few complications around prefetch and has been the reason behind quite a few customer calls.
One example is when prefetching is in progress, customer app closes, re-opens, but does not get messages as they are under "lock" due to the previous in progress prefetch
Current recommendation from the service team is to not implement prefetch now and re-visit it post GA

**Comment by SeanFeldman** [comment]
> We are planning on leaving out Prefetch - it introduces a lot of complexities and has been the cause of many support calls. We can revisit this after GA.

Please don't do this mistake. Claiming that a powerful feature that has complexity when misused should not be exposed at all is equivalent to saying "Service Bus is a complex messaging service that causes many support calls and therefore should be replaced by Storage Queues".

I work with various types of customers. Yes, Prefetch Count, Lock Duration, Competing consumers, Processing time, Lock Auto-renewal are all complex and combined together trip new and inexperienced customers over. That's not a reason to hide the feature that is so **crucial to performance optimization**. Auto-renewal occasionally will fail as it's not a guaranteed operation. Causing message to lose its lock and be reprocessed. Are you going to remove auto-renewal? Lock duration is limited to 5 minutes, are you going to extend that? These are also big contributors to support cases.

> We are planning on leaving out Prefetch

You claim to consult with the community - please do so. Prefetching is a very important feature for customers that have low processing times and value high throughput. Removing access to this feature is literally shooting yourself in afoot.

**Comment by JoshLove-msft** [comment]
@SeanFeldman here is more context as to the decision to leave out Prefetch for GA:
There are quite a few complications around prefetch and has been the reason behind quite a few customer calls. 
One example is when prefetching is in progress, customer app closes, re-opens, but does not get messages as they are under "lock" due to the previous in progress prefetch
Current recommendation from the service team is to not implement prefetch now and re-visit it post GA

**Comment by SeanFeldman** [comment]
> One example is when prefetching is in progress, customer app closes, re-opens, but does not get messages as they are under "lock" due to the previous in-progress prefetch

The maximum lock duration is five minutes. Five. Customers that require to access those messages should be reducing the lock or abandon the messages rather than shutting down. It also depends on how customers design their solutions. Etc, etc. What I'm saying it's much more complicated than just saying "certain percentage of ASB customers don't handle the prefetch correctly". 

The prefetch is disabled by the default (track 1). If it's enabled, that would be the customers that are setting it. I do not find collective punishment the right approach since it will impact all the customers that do it right (or understand how prefetch works and how it relates to concurrency, resources available on the processing endpoints, lock duration, etc).

> Current recommendation from the service team is to not implement prefetch now and re-visit it post GA

While I don't know who has recommended that, I can suggest turning it off at the client level and have an explicit Options if you'd like, but don't hurt the customers that will need it.

**Comment by SeanFeldman** [comment]
>There are quite a few complications around prefetch and has been the reason behind quite a few customer calls.

How many support calls/tickets about LockLostException did Microsoft handle? That one is a much serious headache when it comes to customers trying to design their systems that need to process ASB messages.

**Comment by JoshLove-msft** [comment]
Just to be clear, we are not saying that we will definitely not have Prefetch, but rather that we weren't planning to include for GA, and we would let customer need help to drive whether it is ultimately included.

Yes, we currently have an explicit option to set the `PrefetchCount`, so it is disabled by default. 

Do many of your customers have Prefetch set to greater than 0? This will be useful information in deciding whether/when to add Prefetch support.

**Comment by SeanFeldman** [comment]
Our customers have access to the setting and set it to what they need. By default, we leave it at 0 just as the underlying SDK does.
This feature is more relevant for the customers that prefer throughput over transactional processing (they don't use send-via and can tolerate duplicates). Their single message processing is usually extremely short and the latency of fetching messages is compensated by having a prefetch set to > 0.

> Just to be clear, we are not saying that we will definitely not have Prefetch, but rather that we weren't planning to include for GA, and we would let customer need help to drive whether it is ultimately included.

I understand that. You also have to see it from a different perspective - for years customers were able to leverage prefetch when they needed it. Azure SDK team is coming with a replacement SDK that is supposed to be better than the previous SDKs, encouraging to adopt the new SDK. And yet the SDK will lack the features customers need. This is not a pipeline feature with plugins that were fairly new (about two years from the inception). I'd rather see a fix to a lock-renewal API (not really possible on the SDK side) to fix bigger issues than incorrectly set prefetch count or wrong assumptions in the design.

Long story short, it sounds like the decision has been made. I've brought up the concern, Microsoft can figure out what to do. Thank you for giving me the opportunity to express my concern about this decision.

**Comment by JoshLove-msft** [comment]
Yes, customers were able to leverage prefetch but we don't really have good data on customers who have used it successfully. We do have data on customers that have struggled with it as these end up in support tickets. If you have customers using it successfully, that would be an extremely valuable data point. Again, the decision that was made was to wait until after GA to give us a chance to collect customer feedback - not to say "we are not doing prefetch ever".

Your feedback has been integral to the Track 2 design, so please keep raising any concerns you have.

**Comment by tomkerkhove** [comment]
Let's turn the question around - If the service provides the capability to tweak it, why not give it to customers?

If you ask me, the SDK should allow us to use it; regardless if it's an easy or more complex scenario.

**Comment by tomkerkhove** [comment]
Not having it for GA is ok though, if it's too much work

**Comment by clemensv** [comment]
Prefetch has always been optional and, as Sean describes, is essential for achieving higher throughput on links with higher latencies and while occasional message loss might be okay. 

The argument that some customers might be shooting themselves in the foot when using an optional feature they do no understand is not an argument I buy for cutting said feature. 

We will also introduce service-side changes within the next 6-12 months that will make prefetch significantly safer to use.

I am assuming for the prefetch feature to be staying.

(PS: I'm on vacation until July 13.)

**Comment by andymac4182** [comment]
Prefetch is great when doing large batch jobs. We tune it based on the latency of links and size of work. It can improve the performance of our jobs dramatically. I think it should stay personally.

**Comment by JoshLove-msft** [comment]
Thanks all for the feedback. I will share this with the rest of the feature crew and give an update this week.

**Comment by ramya-rao-a** [comment]
Hey everyone,

A little background on how we got here.

The exercise of working on the new version of the Service Bus SDK is a joint effort across multiple languages(.Net, Java, JS and Python) with the goal of providing the same set of features across the languages, have each language SDK be idiomatic to that language's ecosystem and be consistent with other SDKs in terms of common features like authentication, logging, tracing etc.

While we were looking at implementing prefetch across the four languages, we saw different limitations due to either the platform, or the underlying AMQP library being used. We do not want to go GA with each language having different behavior when it comes to prefetch. While in most languages we were able to come to a consensus on how the prefetch queue should be used (exactly like what we have in the .Net SDK), there were differences on when/how the prefetch queue should be re-populated.

The above along with what @JoshLove-msft has already shared was the reason for the recommendation in the internal team discussions between the SDK and the service team that we take the time to think prefetch through, collect customer feedback (which this issue invariably did a bit :)), see it through the lens of all 4 languages and not rush it. 

For now, we will not make any changes in the .Net SDK and use the time next month to reconcile on how the prefetch feature would look like in the other languages.

Regardless, it was great to hear from all of you on the subject and we hope to keep hearing more as we continue working on this

**Comment by tomkerkhove** [comment]
Can we keep the issue open to track the addition of PrefetchCount at least then?

**Comment by JoshLove-msft** [comment]
> Can we keep the issue open to track the addition of PrefetchCount at least then?

It is already [available](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverOptions.cs#L20).
